File: drills/python/solutions/nc133_lru_cache.py
class LRUCache:
    def __init__(self, capacity):
        # A plain dict, leaning on the fact that Python dicts remember insertion
        # order: deleting a key and putting it back makes it the newest, so the
        # oldest is simply the first key. That is the recency list, free.
        self.capacity = capacity
        self.entries = {}

    def get(self, key):
        if key not in self.entries:
            return -1
        # Reading counts as use, so the key moves to the newest position.
        value = self.entries.pop(key)
        self.entries[key] = value
        return value

    def put(self, key, value):
        if key in self.entries:
            del self.entries[key]
        self.entries[key] = value
        if len(self.entries) > self.capacity:
            oldest = next(iter(self.entries))
            del self.entries[oldest]

File: drills/ts/solutions/nc133_lru_cache.ts
export class LRUCache {
  capacity: number;
  entries: Map<number, number>;

  constructor(capacity: number) {
    // A Map, leaning on the fact that it remembers insertion order: deleting a
    // key and putting it back makes it the newest, so the oldest is simply the
    // first key the iterator yields. That is the recency list, free.
    this.capacity = capacity;
    this.entries = new Map();
  }

  get(key: number): number {
    if (!this.entries.has(key)) return -1;
    // Reading counts as use, so the key moves to the newest position.
    const value = this.entries.get(key)!;
    this.entries.delete(key);
    this.entries.set(key, value);
    return value;
  }

  put(key: number, value: number): void {
    this.entries.delete(key);
    this.entries.set(key, value);
    if (this.entries.size > this.capacity) {
      this.entries.delete(this.entries.keys().next().value!);
    }
  }
}
